pkg/etcdcli/etcdcli_pool_test.go

```diff
@@ -0,0 +1,430 @@
+		client, err := poolRecorder.pool.Get()
+		require.NoError(t, err)
+		assert.NotNil(t, client)
+		clients = append(clients, client)
+	}
+
+	// returning all should make sure the last one tripping over capacity should get closed
+	for _, client := range clients {
+		poolRecorder.pool.Return(client)
+	}
+
+	assert.Equal(t, maxNumCachedClients+1, poolRecorder.numNewCalls)
+	assert.Equal(t, maxNumCachedClients+1, poolRecorder.numEndpointCalls)
+	assert.Equal(t, maxNumCachedClients+1, poolRecorder.numHealthCalls)
+	assert.Equal(t, 1, poolRecorder.numCloseCalls)
+}
+
+func TestNewClientWithTickets(t *testing.T) {
+	integration.BeforeTestExternal(t)
+	testServer := integration.NewClusterV3(t, &integration.ClusterConfig{Size: 3})
+	defer testServer.Terminate(t)
+
+	poolRecorder := newTestPool(testServer)
+	var clients []*clientv3.Client
+	for i := 0; i < maxNumClientTickets; i++ {
+		client, err := poolRecorder.pool.Get()
+		require.NoError(t, err)
+		assert.NotNil(t, client)
+		clients = append(clients, client)
+	}
+
+	assert.Equal(t, maxNumClientTickets, poolRecorder.numNewCalls)
+	assert.Equal(t, maxNumClientTickets, poolRecorder.numEndpointCalls)
+	assert.Equal(t, maxNumClientTickets, poolRecorder.numHealthCalls)
+	assert.Equal(t, 0, poolRecorder.numCloseCalls)
+
+	// this should block and return an error
+	client, err := poolRecorder.pool.Get()
+	assert.Nil(t, client)
+	assert.Errorf(t, err, "too many active cache clients, rejecting to create new one")
+	// returning one should unlock the get again
+	poolRecorder.pool.Return(clients[0])
+	client, err = poolRecorder.pool.Get()
+	assert.NotNil(t, client)
+	assert.Equal(t, maxNumClientTickets, poolRecorder.numNewCalls)        // no new call added
+	assert.Equal(t, maxNumClientTickets+2, poolRecorder.numEndpointCalls) // called Get twice additionally
+	assert.Equal(t, maxNumClientTickets+1, poolRecorder.numHealthCalls)
+	assert.Equal(t, 0, poolRecorder.numCloseCalls)
+}
+
+func TestClosesReturnTickets(t *testing.T) {
+	integration.BeforeTestExternal(t)
+	testServer := integration.NewClusterV3(t, &integration.ClusterConfig{Size: 3})
+	defer testServer.Terminate(t)
+
+	poolRecorder := newTestPool(testServer)
+	var clients []*clientv3.Client
+	for i := 0; i < maxNumClientTickets; i++ {
+		client, err := poolRecorder.pool.Get()
+		require.NoError(t, err)
+		assert.NotNil(t, client)
+		clients = append(clients, client)
+	}
+
+	assert.Equal(t, maxNumClientTickets, poolRecorder.numNewCalls)
+	assert.Equal(t, maxNumClientTickets, poolRecorder.numEndpointCalls)
+	assert.Equal(t, maxNumClientTickets, poolRecorder.numHealthCalls)
+	assert.Equal(t, 0, poolRecorder.numCloseCalls)
+
+	// return all clients to fill the internal cache and cause five to close
+	for i := 0; i < maxNumClientTickets; i++ {
+		poolRecorder.pool.Return(clients[i])
+	}
+	assert.Equal(t, maxNumCachedClients, poolRecorder.numCloseCalls)
+
+	// now we should be able to get the full amount of tickets again
```

Comment:
```suggestion
	// now we should be able to get the full amount of clients again
```
I think you mean clients and not tickets here as the number of available tickets is just 5 at this point.
This threw me off initially as it made me think the tickets were returned once the client was as well.

So I expected the following unit test to succeed.
```Go
func TestAllTicketsAvailableForFullyCachedPool(t *testing.T) {
	integration.BeforeTestExternal(t)
	testServer := integration.NewClusterV3(t, &integration.ClusterConfig{Size: 3})
	defer testServer.Terminate(t)

	poolRecorder := newTestPool(testServer)
	var clients []*clientv3.Client
	for i := 0; i < maxNumCachedClients; i++ {
		client, err := poolRecorder.pool.Get()
		require.NoError(t, err)
		assert.NotNil(t, client)
		clients = append(clients, client)
	}

	// return all clients to the client pool and check that none are closed
	for _, client := range clients {
		poolRecorder.pool.Return(client)
	}
	assert.Equal(t, maxNumCachedClients, poolRecorder.numNewCalls)
	assert.Equal(t, maxNumCachedClients, poolRecorder.numEndpointCalls)
	assert.Equal(t, maxNumCachedClients, poolRecorder.numHealthCalls)
	assert.Equal(t, 0, poolRecorder.numCloseCalls)

	// The tickets should equal the max available (10) since all clients are ready to be taken from the cached pool
	assert.Equal(t, maxNumClientTickets, len(poolRecorder.pool.availableTickets))

	// Grab a cached client from the pool
	client, err := poolRecorder.pool.Get()
	require.NoError(t, err)
	assert.NotNil(t, client)
	assert.Equal(t, maxNumClientTickets-1, len(poolRecorder.pool.availableTickets))

	// Return it back and all tickets should be available again
	poolRecorder.pool.Return(client)
	assert.Equal(t, maxNumClientTickets, len(poolRecorder.pool.availableTickets))
}
```

But looking through the code it seems tickets used == number of clients currently open. So just wanted to confirm this.

Reply:
indeed, there are two limits:
* pool size, how many active clients we are keeping cached at max
* tickets, how many clients we are handing out at a time

The first should ensure we are not creating new clients all the time and keeping them cached, because the TLS handshakes are expensive. Yet bound to `maxNumCachedClients` as too many IDLE connections to etcd are also not helpful.

The second should ensure that on connectivity issues we do not overwhelm etcd additionally by creating more connections. The behavior in a previous version before the ticketing was: 
* I have to get the members, can I get a cached client?
* pool is filled, let me grab one and try to health check
* none of the healthy client are available, let's create a new one

The last bullet was basically creating hundreds of new clients as there was no connectivity for some time period. The ticketing basically rate limits this, imagine going to a full amusement park and you have to wait for somebody to come out of it. 

> But looking through the code it seems tickets used == number of clients currently open. So just wanted to confirm this.

yep, good summary. Shall we rename that ticketing aspect to `numOpenClients`?